drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,713 @@
+#include <errno.h>
+#include <gpio.h>
+#include <i2c.h>
+#include <misc/util.h>
+#include <string.h>
+
+/* Number of entries in hardware command queue */
+#define I2C_ESP32_NUM_CMDS 16
+/* Number of bytes in hardware FIFO */
+#define I2C_ESP32_BUFFER_SIZE 32
+
+#define I2C_ESP32_TIMEOUT_MS 100
+#define I2C_ESP32_SPIN_THRESHOLD 10
+#define I2C_ESP32_YIELD_THRESHOLD (I2C_ESP32_SPIN_THRESHOLD / 2)
+#define I2C_ESP32_TIMEOUT \
+	((I2C_ESP32_YIELD_THRESHOLD) + (I2C_ESP32_SPIN_THRESHOLD))
+
+enum i2c_esp32_opcodes {
+	I2C_ESP32_OP_RSTART,
+	I2C_ESP32_OP_WRITE,
+	I2C_ESP32_OP_READ,
+	I2C_ESP32_OP_STOP,
+	I2C_ESP32_OP_END
+};
+
+struct i2c_esp32_cmd {
+	u32_t num_bytes : 8;
+	u32_t ack_en : 1;
+	u32_t ack_exp : 1;
+	u32_t ack_val : 1;
+	u32_t opcode : 3;
+	u32_t : 17;
+	u32_t done : 1;
+};
+
+struct i2c_esp32_data {
+	u32_t dev_config;
+	u16_t address;
+
+	struct k_sem fifo_sem;
+	struct k_sem transfer_sem;
+};
+
+typedef void (*irq_connect_cb)(void);
+
+struct i2c_esp32_config {
+	int index;
+
+	irq_connect_cb connect_irq;
+
+	const struct {
+		int sda_out;
+		int sda_in;
+		int scl_out;
+		int scl_in;
+	} sig;
+
+	const struct {
+		int scl;
+		int sda;
+	} pins;
+
+	const struct {
+		int clk;
+		int rst;
+	} enable_bits;
+
+	const struct {
+		bool tx_lsb_first;
+		bool rx_lsb_first;
+	} mode;
+
+	const struct {
+		int source;
+		int line;
+	} irq;
+
+	const u32_t default_config;
+};
+
+static inline void set_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) | v, mem_addr);
+}
+
+static inline void clear_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) & ~v, mem_addr);
+}
+
+static void i2c_esp32_enable_peripheral(const struct i2c_esp32_config *config)
+{
+	set_mask32(config->enable_bits.clk, DPORT_PERIP_CLK_EN_REG);
+	clear_mask32(config->enable_bits.rst, DPORT_PERIP_RST_EN_REG);
+}
+
+static const char *i2c_esp32_get_gpio_for_pin(int pin)
+{
+	if (pin < 32) {
+#if defined(CONFIG_GPIO_ESP32_0)
+		return CONFIG_GPIO_ESP32_0_NAME;
+#else
+		return NULL;
+#endif /* CONFIG_GPIO_ESP32_0 */
+	}
+
+#if defined(CONFIG_GPIO_ESP32_1)
+	return CONFIG_GPIO_ESP32_1_NAME;
+#else
+	return NULL;
+#endif /* CONFIG_GPIO_ESP32_1 */
+}
+
+static int i2c_esp32_configure_pins(int pin, int matrix_out, int matrix_in)
+{
+	const int pin_mode = GPIO_DIR_OUT |
+			     GPIO_DS_DISCONNECT_LOW |
+			     GPIO_PUD_PULL_UP;
+	const char *device_name = i2c_esp32_get_gpio_for_pin(pin);
+	struct device *gpio;
+	int ret;
+
+	if (!device_name) {
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	ret = gpio_pin_configure(gpio, pin, pin_mode);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = gpio_pin_write(gpio, pin, 1);
+	if (ret < 0) {
+		return ret;
+	}
+
+	esp32_rom_gpio_matrix_out(pin, matrix_out, false, false);
+	esp32_rom_gpio_matrix_in(pin, matrix_in, false);
+
+	return 0;
+}
+
+static int i2c_esp32_configure_frequency(const struct i2c_esp32_config *config,
+					 u32_t speed)
```

Comment:
Style whine: function name says "frequency", parameter name is "speed", and what it actually is is an enumerant.  Trying to pass this thing a frequency number in khz or whatever is going to generate a giant buffer overrun.   Really the i2c layer should have defined these as an enum to begin with I  guess.  Maybe rename the parameter "speed_enum" or something for clarity if you do another spin?